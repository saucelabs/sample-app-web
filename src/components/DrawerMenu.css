@media only screen and (max-width: 480px) {
  .bm-menu-wrap {
    width: 100% !important;
  }
}
.menu-item {
  border-bottom: 2px solid #474c55;
  color: #474c55;
  display: inline-block;
  font-weight: 400;
  margin-bottom: 7px;
  text-decoration: none;
  cursor: pointer;
  text-transform: uppercase;
  padding: 10px 0;
}

.menu-item:hover {
  border-bottom: 2px solid #e2231a;
  color: #e2231a;
}

/* ===============Burger button styling==================== */
/* Position and sizing of burger button */
.bm-burger-button {
  position: absolute;
  width: 30px;
  height: 30px;
  left: 15px;
  top: 20px;
}

/* Color/shape of burger icon bars */
.bm-burger-bars {
  background: #777;
}

/* Position and sizing of clickable cross button */
.bm-cross-button {
  height: 40px !important;
  width: 40px !important;
}

/* Color/shape of close button cross */
.bm-cross {
  background: #ffffff;
}

/* General sidebar styles */
.bm-menu {
  background: #fff;
  padding: 2.5em 1.5em 0;
  font-size: 16px;
  box-shadow: none;
}

/* Morph shape necessary with bubble or elastic */
.bm-morph-shape {
  fill: #373a47;
}

/* Individual item */
.bm-item {
}
