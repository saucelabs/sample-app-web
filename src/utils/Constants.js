export const FONT_FAMILY = "Roboto, Arial, Helvetica, sans-serif";
export const VALID_USERNAMES = [
  "standard_user",
  "locked_out_user",
  "problem_user",
  "performance_glitch_user",
  "error_user",
  "visual_user",
];
export const VALID_PASSWORD = "secret_sauce";
export const ROUTES = {
  LOGIN: "/",
  INVENTORY: "/inventory.html",
  INVENTORY_LIST: "/inventory-item.html",
  CART: "/cart.html",
  CHECKOUT_STEP_ONE: "/checkout-step-one.html",
  CHECKOUT_STEP_TWO: "/checkout-step-two.html",
  CHECKOUT_COMPLETE: "/checkout-complete.html",
};
export const CART_CONTENTS = "cart-contents";
export const SESSION_USERNAME = "session-username";
