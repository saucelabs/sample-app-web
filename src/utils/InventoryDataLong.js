export const InventoryDataLong = [
  {
    id: 0,
    name: "Sauce Labs Bike Light",
    desc: `A red light isn't the desired state in testing but it sure helps when riding your bike at night. Water-resistant with 3 lighting modes, 1 AAA battery included.`,
    price: 9.99,
    image_url: "bike-light-1200x1500.jpg",
  },
  {
    id: 1,
    name: "Sauce Labs Bolt T-Shirt (XS)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 2,
    name: "Sauce Labs Onesie",
    desc: `Rib snap infant onesie for the junior automation engineer in development. Reinforced 3-snap bottom closure, two-needle hemmed sleeved and bottom won't unravel.`,
    price: 7.99,
    image_url: "red-onesie-1200x1500.jpg",
  },
  {
    id: 3,
    name: "Test.allTheThings() T-Shirt (Red) (XS)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },
  {
    id: 4,
    name: "Sauce Labs Backpack",
    desc: `carry.allTheThings() with the sleek, streamlined Sly Pack that melds uncompromising style with unequaled laptop and tablet protection.`,
    price: 29.99,
    image_url: "sauce-backpack-1200x1500.jpg",
  },
  {
    id: 5,
    name: "Sauce Labs Fleece Jacket (XS)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 6,
    name: "Sauce Labs Fleece Jacket (S)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 7,
    name: "Sauce Labs Fleece Jacket (M)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 8,
    name: "Sauce Labs Fleece Jacket (L)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 9,
    name: "Sauce Labs Fleece Jacket (XL)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 10,
    name: "Sauce Labs Fleece Jacket (XXL)",
    desc: `It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.`,
    price: 49.99,
    image_url: "sauce-pullover-1200x1500.jpg",
  },
  {
    id: 11,
    name: "Test.allTheThings() T-Shirt (Red) (S)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 12,
    name: "Test.allTheThings() T-Shirt (Red) (M)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 13,
    name: "Test.allTheThings() T-Shirt (Red) (L)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 14,
    name: "Test.allTheThings() T-Shirt (Red) (L)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 15,
    name: "Test.allTheThings() T-Shirt (Red) (XL)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },

  {
    id: 16,
    name: "Test.allTheThings() T-Shirt (Red) (XXL)",
    desc: `This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.`,
    price: 15.99,
    image_url: "red-tatt-1200x1500.jpg",
  },
  {
    id: 17,
    name: "Sauce Labs Bolt T-Shirt (M)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 18,
    name: "Sauce Labs Bolt T-Shirt (L)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 19,
    name: "Sauce Labs Bolt T-Shirt (XL)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 20,
    name: "Sauce Labs Bolt T-Shirt (XXL)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
  {
    id: 21,
    name: "Sauce Labs Bolt T-Shirt (S)",
    desc: `Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.`,
    price: 15.99,
    image_url: "bolt-shirt-1200x1500.jpg",
  },
];
