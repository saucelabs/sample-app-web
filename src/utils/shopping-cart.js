export class ShoppingCart {
  static addItem(itemId) {
    // pull out our current cart contents
    const curContents = ShoppingCart.getCartContents();

    /* istanbul ignore else */
    if (curContents.indexOf(itemId) < 0) {
      // Item's not yet present - add it now
      curContents.push(itemId);

      // We modified our cart, so store it now
      ShoppingCart.setCartContents(curContents);
    }
  }

  static removeItem(itemId) {
    // pull out our current cart contents
    const curContents = ShoppingCart.getCartContents();
    const itemIndex = curContents.indexOf(itemId);

    /* istanbul ignore else */
    if (itemIndex >= 0) {
      // Remove this item from the array
      curContents.splice(itemIndex, 1);

      // We modified our cart, so store it now
      ShoppingCart.setCartContents(curContents);
    }
  }

  static isItemInCart(itemId) {
    // pull out our current cart contents
    const curContents = ShoppingCart.getCartContents();

    // If the item is in the array, return true
    return curContents.indexOf(itemId) >= 0;
  }

  static getCartContents() {
    // pull out our current cart contents
    let curContents = window.localStorage.getItem("cart-contents");

    // Make an empty list if this is the first item
    if (curContents == null) {
      curContents = [];
    } else {
      // We have an existing cart, so deserialize it now since localStorage stores in JSON strings
      curContents = JSON.parse(curContents);
    }

    return curContents;
  }

  static setCartContents(newContents) {
    window.localStorage.setItem("cart-contents", JSON.stringify(newContents));

    // Notify our listeners
    /* istanbul ignore next */
    ShoppingCart.LISTENERS.forEach((curListener) => {
      curListener.forceUpdate();
    });
  }

  static resetCart() {
    window.localStorage.removeItem("cart-contents");

    // Notify our listeners
    /* istanbul ignore next */
    ShoppingCart.LISTENERS.forEach((curListener) => {
      curListener.forceUpdate();
    });
  }

  static registerCartListener(handler) {
    ShoppingCart.LISTENERS.push(handler);
  }
}

ShoppingCart.LISTENERS = [];
