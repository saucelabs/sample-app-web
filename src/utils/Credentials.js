import Cookies from "js-cookie";
import { SESSION_USERNAME, VALID_PASSWORD, VALID_USERNAMES } from "./Constants";

/**
 * Verify the credentials
 *
 * @param {string} username
 * @param {string} password
 *
 * @return {boolean}
 */
export function verifyCredentials(username, password) {
  if (password !== VALID_PASSWORD) {
    return false;
  }

  return VALID_USERNAMES.includes(username);
}

/**
 * Store the data in our cookies
 *
 * @param {string} username
 *
 * @param {string} password
 */
export function setCredentials(username, password) {
  let date = new Date();
  date.setTime(date.getTime() + 10 * 60 * 1000);

  Cookies.set(SESSION_USERNAME, username, { expires: date });
}

/**
 * Remove the credentials
 */
export function removeCredentials() {
  Cookies.remove(SESSION_USERNAME);
}

/**
 * Return current logged username
 *
 * @return {string | undefined}
 */
export function currentUser() {
  return Cookies.get(SESSION_USERNAME);
}

/**
 * Check if this is a problem user
 *
 * @return {boolean}
 */
export function isProblemUser() {
  return Cookies.get(SESSION_USERNAME) === "problem_user";
}

/**
 * Check if this is a performance user
 *
 * @return {boolean}
 */
export function isPerformanceGlitchUser() {
  return Cookies.get(SESSION_USERNAME) === "performance_glitch_user";
}

/**
 * Check if this a logged out user
 *
 * @return {boolean}
 */
export function isLockedOutUser() {
  return Cookies.get(SESSION_USERNAME) === "locked_out_user";
}

/**
 * Check if this is an error user
 *
 * @return {boolean}
 */
export function isErrorUser() {
  return Cookies.get(SESSION_USERNAME) === "error_user";
}

/**
 * Check if the user is logged in with a valid username
 *
 * @return {boolean}
 */
export function isLoggedIn() {
  const sessionUsername = Cookies.get(SESSION_USERNAME);
  const isValidUsername = VALID_USERNAMES.includes(sessionUsername);

  return isValidUsername && sessionUsername !== "locked_out_user";
}

/**
 * Check if this is a visual user
 *
 * @return {boolean}
 */
export function isVisualUser() {
  return Cookies.get(SESSION_USERNAME) === "visual_user";
}
