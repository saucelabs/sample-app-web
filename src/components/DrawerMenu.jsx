import React from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { slide as Menu } from "react-burger-menu";
import { ShoppingCart } from "../utils/shopping-cart";
import { ROUTES } from "../utils/Constants";
import { isProblemUser, removeCredentials } from "../utils/Credentials";
import menuClosePng from "../assets/img/close.png";
import menuCloseSvg from "../assets/svg/close@3x.svg";
import menuIconPng from "../assets/img/menu.png";
import menuIconSvg from "../assets/svg/menu3x.svg";
import "./DrawerMenu.css";

const DrawerMenu = ({ history }) => {
  const resetStorage = () => {
    // Wipe out our shopping cart now
    ShoppingCart.resetCart();
  };
  const aboutLink = isProblemUser()
    ? "https://saucelabs.com/error/404"
    : "https://saucelabs.com/";

  return (
    <Menu
      customBurgerIcon={
        <img src={menuIconPng} srcSet={menuIconSvg} alt="Open Menu" />
      }
      customCrossIcon={
        <img src={menuClosePng} srcSet={menuCloseSvg} alt="Close Menu" />
      }
      outerContainerId={"page_wrapper"}
      pageWrapId={"contents_wrapper"}
      noOverlay
    >
      <a
        id="inventory_sidebar_link"
        className="menu-item"
        href="#"
        onClick={(evt) => {
          evt.preventDefault();
          history.push(ROUTES.INVENTORY);
        }}
      >
        All Items
      </a>
      <a id="about_sidebar_link" className="menu-item" href={aboutLink}>
        About
      </a>
      <a
        id="logout_sidebar_link"
        className="menu-item"
        href="#"
        onClick={(evt) => {
          evt.preventDefault();
          removeCredentials();
          history.push(ROUTES.LOGIN);
        }}
      >
        Logout
      </a>
      <a
        id="reset_sidebar_link"
        className="menu-item"
        href="#"
        onClick={(evt) => {
          evt.preventDefault();
          resetStorage();
        }}
      >
        Reset App State
      </a>
    </Menu>
  );
};
DrawerMenu.propTypes = {
  /**
   * The history
   */
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

export default withRouter(DrawerMenu);
