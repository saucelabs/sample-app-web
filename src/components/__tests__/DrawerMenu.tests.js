import React from "react";
import { shallow } from "enzyme";
import DrawerMenu from "../DrawerMenu";
import * as Credentials from "../../utils/Credentials";
import { ShoppingCart } from "../../utils/shopping-cart";

jest.mock("../../utils/shopping-cart");

let props;

describe("DrawerMenu", () => {
  beforeEach(() => {
    props = {
      history: {
        push: jest.fn(),
      },
    };
  });

  it("should render correctly", () => {
    const component = shallow(<DrawerMenu.WrappedComponent {...props} />);

    expect(component).toMatchSnapshot();
  });

  it("should render return an incorrect about link for a problem user", () => {
    const isProblemUserSpy = jest.spyOn(Credentials, "isProblemUser");
    isProblemUserSpy.mockReturnValue(true);
    const component = shallow(<DrawerMenu.WrappedComponent {...props} />);
    const aboutLink = component.find("a").at(1);
    aboutLink.simulate("click", {
      preventDefault() {},
    });

    expect(isProblemUserSpy).toHaveBeenCalledTimes(1);
    expect(aboutLink.props().href).toEqual("https://saucelabs.com/error/404");

    isProblemUserSpy.mockClear();
  });

  it("should be able to redirect to the inventory page when clicking on the all items link", () => {
    const component = shallow(<DrawerMenu.WrappedComponent {...props} />);
    const allItemsLink = component.find("a").at(0);
    allItemsLink.simulate("click", {
      preventDefault() {},
    });

    expect(props.history.push).toHaveBeenCalledWith("/inventory.html");
  });

  it("should be able to redirect to the login page when clicking on the logout link", () => {
    const component = shallow(<DrawerMenu.WrappedComponent {...props} />);
    const logoutLink = component.find("a").at(2);
    const removeCredentialsSpy = jest.spyOn(Credentials, "removeCredentials");
    removeCredentialsSpy.mockReturnValue(true);
    logoutLink.simulate("click", {
      preventDefault() {},
    });

    expect(props.history.push).toHaveBeenCalledWith("/");
    expect(removeCredentialsSpy).toHaveBeenCalledTimes(1);
  });

  it("should be able to reset the storage when Reset App State is being called", () => {
    const component = shallow(<DrawerMenu.WrappedComponent {...props} />);
    const resetAppStateLink = component.find("a").at(3);
    ShoppingCart.resetCart = jest.fn();
    resetAppStateLink.simulate("click", {
      preventDefault() {},
    });

    expect(ShoppingCart.resetCart).toHaveBeenCalledTimes(1);
  });
});
