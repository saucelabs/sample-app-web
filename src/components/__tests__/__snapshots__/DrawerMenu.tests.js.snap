// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerMenu should render correctly 1`] = `
<Menu
  bodyClassName=""
  burgerBarClassName=""
  burgerButtonClassName=""
  className=""
  crossButtonClassName=""
  crossClassName=""
  customBurgerIcon={
    <img
      alt="Open Menu"
      src="test-file-stub"
      srcSet="test-file-stub"
    />
  }
  customCrossIcon={
    <img
      alt="Close Menu"
      src="test-file-stub"
      srcSet="test-file-stub"
    />
  }
  disableAutoFocus={false}
  disableCloseOnEsc={false}
  htmlClassName=""
  id=""
  itemClassName=""
  itemListClassName=""
  itemListElement="nav"
  menuClassName=""
  morphShapeClassName=""
  noOverlay={true}
  noTransition={false}
  onIconHoverChange={[Function]}
  onStateChange={[Function]}
  outerContainerId="page_wrapper"
  overlayClassName=""
  pageWrapId="contents_wrapper"
  styles={Object {}}
  width={300}
>
  <a
    className="menu-item"
    href="#"
    id="inventory_sidebar_link"
    onClick={[Function]}
  >
    All Items
  </a>
  <a
    className="menu-item"
    href="https://saucelabs.com/"
    id="about_sidebar_link"
  >
    About
  </a>
  <a
    className="menu-item"
    href="#"
    id="logout_sidebar_link"
    onClick={[Function]}
  >
    Logout
  </a>
  <a
    className="menu-item"
    href="#"
    id="reset_sidebar_link"
    onClick={[Function]}
  >
    Reset App State
  </a>
</Menu>
`;
